Fix file upload EOS bug.

bgroenks96 opened the pull request:
Per the .NET Stream API, Read returns zero ONLY on end-of-stream.

Thus, the file upload should not continue after the inner read/write
loop terminates without error.

Stream.Length should not be relied upon for this purpose, as many
implementations of Stream may be buffered, and there is no guarantee
that the length of the Stream is the size of the file.

Resolves #174

artiomchi commented:
Good PR, but could you make sure your code uses the same whitespace character as the project? You're using spaces, and the project uses tabs, which makes reading the diff a big pain :)

In reality, this PR just added 4 `break;` calls, but it looks like the change was MUCH more extensive

artiomchi commented:
Could you apply the same fix to the `DownloadInternal` and `DownloadInternalAsync` methods in the same way? They have the same problem (with a poor attempt to overcome it).

I'm just testing a private build of the library where I removed the `readToEnd` variable entirely, and replaced it with `break` statements, same as in your code. Hoping it'll fix the issue we're seeing.

I'd create a PR of my own, but I just did the same thing you did, so want to give proper credit to you :)

bgroenks96 commented:
@artiomchi Absolutely! I'll make the changes as soon as I am able.

bgroenks96 commented:
@artiomchi I cleaned up the commit to remove some of the whitespace changes you referred to. I hope that makes it clearer.

Unfortunately, after further investigation, I don't think adding the unconditional `break` to file _downloads_ is a good idea (and also isn't likely the source of your problem).

Unlike the upload case, we _know_ the expected file length as reported by the server. The `readToEnd` flag correctly covers the case under which me must rely on the EOS signal to finish the download.

That being said, it's still an error to continue to read from the stream after it returns zero. I have modified the download methods to throw an exception and attempt to resume the download in this case. I would be interested to see if that solves your issue.

Best of luck!

artiomchi commented:
@bgroenks96 ah, I see, that's a great observation! I'll apply this patch instead of mine then. Thanks for that!

Unfortunately, I learned that a lot of the FTP servers that we have to connect to are pretty bad. To the point that `AutoActive` and `AutoPassive` modes don't work (server and client negotiate to a newer transfer mode, that the server reports as being supported, but it really doesn't), so I have to force it to use `PASV` and `PORT` instead. At this point I wouldn't be surprised if they sent the wrong file size as well :)

hgupta9 commented:
Thank you very much @artiomchi and @bgroenks96 for this fix! Much appreciated!
